File: apps/web/src/types/portfolio.ts
import type { HoldingRead } from "./holding";

export interface PortfolioRead {
  id: number;
  name: string;
  created_at: string;
  updated_at: string;
}

export interface PortfolioDetail extends PortfolioRead {
  holdings: HoldingRead[];
}

export interface PortfolioCreate {
  name: string;
}

export interface PortfolioUpdate {
  name?: string;
}

File: apps/api/app/schemas/portfolio.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict

from .holding import HoldingRead


class PortfolioCreate(BaseModel):
    name: str


class PortfolioUpdate(BaseModel):
    name: str | None = None


class PortfolioRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    created_at: datetime
    updated_at: datetime


class PortfolioDetail(PortfolioRead):
    """Portfolio with nested holdings list."""

    holdings: list[HoldingRead] = []
